---
title: "In silico test"
author:
  - name: Yan Hui
    affiliation: Department of Food Science, University of Copenhagen
    email: huiyan@food.ku.dk
date: 17/10/2023
output:
  html_document:
    toc: true
link-citations: yes
bibliography: ref.bib
csl: cell.csl
---

```{r, meshclust_threshold}
# recall, percentage of reads correctly being grouped into one cluster
# loss, percentage of reads not being grouped into any cluster
library(data.table)
library(tidyverse)
# fqs_header: TP
fs <- list.files(
  path = "./data/simulate/clust/fqs_header",
  pattern = "*[0-9].txt", full.names = TRUE, recursive = TRUE)
dt <- rbindlist(sapply(fs, function(x) fread(x, header=FALSE, sep=""), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt$benchmark <- gsub("^.*fqs_header/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)_(even10|skew10)_si.*.txt", "\\1", dt$file) #nolint
dt$profile <- gsub("^.*fqs_header/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)_(even10|skew10)_si.*.txt", "\\2", dt$file) #nolint
dt$id <- as.numeric(gsub("^.*fqs_header/.*_si(9095|9597|9798|9899|99100)([0-9]+).txt", "\\1", dt$file)) #nolint
dt$depth <- as.numeric(gsub("^.*fqs_header/.*_si(9095|9597|9798|9899|99100)([0-9]+).txt", "\\2", dt$file)) #nolint
dt <- dt[, !"file"]
# everthing before the first whitespace
dt$seqid <- sub(" .*$", "", dt$V1)
dt$seqid <- sub("^@", "", dt$seqid)
dt$len <- as.numeric(gsub("^.*\ length=([0-9]+)\ .*", "\\1", dt$V1))
dt$eflen <- as.numeric(gsub("^.*error-free_length=([0-9]+).*", "\\1", dt$V1))
dt$id2ref <- as.numeric(gsub("^.*read_identity=(.*)%", "\\1", dt$V1))

# separate V2 by ',' into multiple columns
dt$V2 <- gsub("[^ ]+ (.*) length=.*", "\\1", dt$V1)
# adaptor has no direction, but simulted strand does
dt <- dt[, c("V3", "V4", "V5", "V6", "V7") := tstrsplit(V2, " ", fixed = TRUE)] #nolint
dt <- dt[, c("ref", "strand", "pos") := tstrsplit(V3, ",", fixed = TRUE)] #nolint
dt$type <- dt$ref
dt$type[dt$V4 == "chimera"] <- "chimera"
# ref_[0-9]+ -> ref_seq
dt$type <- gsub("^ref_[0-9]+", "ref_seq", dt$type)
dt <- dt[, !c("V1", "V3", "V4", "V5", "V6", "V7")]
df_ref <- as.data.frame(dt)
colnames(df_ref) <- c("benchmark", "profile", "id", "depth", "seqid", "len", "eflen", "id2ref", "info", "ref", "strand", "pos", "type") #nolint
df_ref$ref_strand <- paste0(df_ref$ref, df_ref$strand)

# cluster: P
fs <- list.files(
  path = "./data/simulate/clust",
  pattern = "*.tsv", full.names = TRUE, recursive = TRUE)
# grep meshclust_..
fs <- grep(pattern = '.*/meshclust_.*', fs, value = TRUE)
dt <- rbindlist(sapply(fs, function(x) fread(x, header=FALSE, sep="\t", blank.lines.skip = TRUE), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt$benchmark <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\1", dt$file) #nolint
dt$profile <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\2", dt$file) #nolint
dt$meshclust <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*meshclust_(auto|70|75|80|85|90|95)/.*", "\\3", dt$file) #nolint
dt$id <- as.numeric(gsub("^.*/si(9095|9597|9798|9899|99100).*", "\\1", dt$file)) #nolint
dt$depth <- as.numeric(
  gsub("^.*/si(9095|9597|9798|9899|99100)([0-9]+).*", "\\2", dt$file)) #nolint
# rm ">" in V2
dt$V2 <- gsub("^>", "", dt$V2)
dt <- dt[, !"file"]
df_cls <- as.data.frame(dt)
colnames(df_cls) <- c("cluster", "seqid", "est_id", "member", "benchmark", "profile", "meshclust", "id", "depth") #nolint
df_merge <- df_cls %>%
  full_join(df_ref, by = c("seqid", "benchmark", "profile", "id", "depth")) %>%
  group_by(benchmark, profile, meshclust, id, depth, cluster) %>%
  # number of reads supporting the cluster
  mutate(snreads = length(unique(seqid))) %>%
  ungroup() %>%
  # number of reads supporting the class
  group_by(benchmark, profile, meshclust, id, depth, cluster, ref_strand) %>%
  mutate(snreads2 = length(unique(seqid))) %>%
  # sort by nsreads, the most common class comes first
  arrange(desc(snreads2), ref_strand, .by_group = TRUE) %>%
  ungroup()

# snreads >= 10
df_merge <- df_merge %>%
  filter(snreads >= 10)

#https://stackoverflow.com/questions/9253843/r-clustering-purity-metric
purity <- function(clusters, classes) {
  sum(apply(table(classes, clusters), 2, max)) / length(clusters)
}
library(aricode)

df_comp0 <- df_merge %>%
  filter(! is.na(cluster)) %>%
  group_by(benchmark, profile, meshclust, id, depth, cluster, snreads) %>%
  summarize(
    Purity = purity(as.integer(as.ordered(cluster)), as.integer(as.ordered(ref_strand))), #nolint
  ) %>%
  ungroup() %>%
  gather(key, value, -benchmark, -profile, -meshclust, -id, -depth, -cluster, -snreads) #nolint
library(ggprism)
library(RColorBrewer)
p0_cls <- df_comp0 %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS"))) %>%
  ggplot( 
    aes(
      x = snreads,
      y = value, 
      color = factor(
        meshclust,
        levels = c("auto", "70", "75", "80", "85", "90", "95"), #nolint
        labels = c("auto", "0.7", "0.75", "0.8", "0.85", "0.9", "0.95") #nolint
        ),
      shape = factor(
        profile,
        levels = c("even10", "skew10")
        ))) +  
  geom_point() +
  labs(
    x = "# of supporrting reads",
    y = "Purity",
    color = "Meshclust threshold",
    shape = "Profile"
   ) +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,2,3,4,5,7,8)]) +
  facet_grid(
    benchmark ~ factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ), scales = "free") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    )
#p0_cls

df_comp <- df_merge %>%
  filter(! is.na(cluster)) %>%
  group_by(benchmark, profile, meshclust, id, depth) %>%
  summarise(
    Size = length(unique(cluster)),
    Purity = purity(as.integer(as.ordered(cluster)), as.integer(as.ordered(ref_strand))), #nolint
    NMI = NMI(as.integer(as.ordered(ref_strand)), as.integer(as.ordered(cluster))), #nolint
  ) %>%
  gather(key, value, -benchmark, -profile, -meshclust, -id, -depth) #nolint

p1_cls <- df_comp %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS"))) %>%
  ggplot( 
    aes(
      x = factor(
        meshclust,
        levels = c("auto", "70", "75", "80", "85", "90", "95"), #nolint
        labels = c("auto", "0.7", "0.75", "0.8", "0.85", "0.9", "0.95") #nolint
        ),
      y = value, 
      color = factor(
        id, 
        levels = c(9095,9597,9798,9899,99100),
        labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%")), #nolint
      fill = factor(
        id, 
        levels = c(9095,9597,9798,9899,99100),
        labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
        ))) +
  geom_col(position = position_dodge2(width = 0.8, preserve = "single")) + #nolint
  geom_hline(data = data.frame(yint=c(20,1,1),key=c("Size", "Purity", "NMI")), aes(yintercept = yint), linetype = "dashed") + #nolint
  labs(
    x = "Meshclust threshold",
    y = "Quality of clustering",
    color = "% of Divergence",
    fill = "% of Divergence"
   ) +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  scale_fill_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  facet_grid(
    factor(
      key,
      levels = c("Size", "Purity", "NMI")) ~ benchmark + factor(
      profile,
      levels = c("even10", "skew10")), #nolint
    scales = "free", space = "free_x") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    strip.text = element_text(size = 8),
    )
#p1_cls
# pct of clustered reads
df_clust <- df_merge %>%
  group_by(benchmark, profile, id, depth) %>%
  mutate(nreads = length(unique(seqid))) %>%
  filter(! is.na(cluster)) %>%
  group_by(benchmark, profile, meshclust, id, depth, type) %>%
  reframe(pct = length(unique(seqid))/nreads * 100)  %>%
  # rm duplicate lines
  distinct()

p1_clust <- df_clust %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS"))) %>%
  ggplot( 
    aes(
      x = factor(
        meshclust,
        levels = c("auto", "70", "75", "80", "85", "90", "95"), #nolint
        labels = c("auto", "0.7", "0.75", "0.8", "0.85", "0.9", "0.95") #nolint
        ),
      y = pct, 
        )) +
  geom_bar(stat = "identity", aes(fill = type)) +
  labs(
    x = "Meshclust threshold",
    y = "% of clustered reads",
    fill = "Type"
   ) +
  scale_fill_manual(values = brewer.pal(4, "Spectral")[c(1,4)]) +
  facet_grid(
    factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ) ~ benchmark + profile,
      scales = "free_x", space = "free_x") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    # axis font angle
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    strip.text = element_text(size = 8),
    )
#p1_clust
p1_clustid <- df_merge %>%
  filter(! is.na(cluster)) %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS"))) %>%
  filter(!is.na(cluster)) %>%
  ggplot( 
    aes(
      x = factor(
        meshclust,
        levels = c("auto", "70", "75", "80", "85", "90", "95"), #nolint
        labels = c("auto", "0.7", "0.75", "0.8", "0.85", "0.9", "0.95") #nolint
        ),
      y = id2ref, 
        )) +
  geom_boxplot(
    aes(fill = type),
    position = position_dodge(preserve = "single"),
    outlier.size = 0.1,
    ) +
  labs(
    x = "Meshclust threshold",
    y = "% of identity relative to reference",
    fill = "Type"
   ) +
  scale_fill_manual(values = brewer.pal(4, "Spectral")[c(1,4)]) +
  facet_grid(
    factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ) ~ benchmark + profile,
      scales = "free_x", space = "free_x") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    # axis font angle
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    strip.text = element_text(size = 8)
    )
#p1_clustid

dir.create("./figure/simulate/meshclust_threshold", recursive = TRUE, showWarnings = FALSE) #nolint
ggsave("./figure/simulate/meshclust_threshold/p0_cls_point.png", p0_cls, width = 12, height = 10) #nolint
ggsave("./figure/simulate/meshclust_threshold/p1_cls_bar.png", p1_cls, width = 12, height = 8) #nolint
ggsave("./figure/simulate/meshclust_threshold/p1_clust_pct.png", p1_clust, width = 12, height = 12) #nolint
ggsave("./figure/simulate/meshclust_threshold/p1_clustid_identity.png", p1_clustid, width = 12, height = 12) #nolint
```

```{r, clust_mode}
# recall, percentage of reads correctly being grouped into one cluster
# loss, percentage of reads not being grouped into any cluster
library(data.table)
library(tidyverse)
# fqs_header: TP
fs <- list.files(
  path = "./data/simulate/clust/fqs_header",
  pattern = "*[0-9].txt", full.names = TRUE, recursive = TRUE)
dt <- rbindlist(sapply(fs, function(x) fread(x, header=FALSE, sep=""), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt$benchmark <- gsub("^.*fqs_header/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)_(even10|skew10)_si.*.txt", "\\1", dt$file) #nolint
dt$profile <- gsub("^.*fqs_header/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)_(even10|skew10)_si.*.txt", "\\2", dt$file) #nolint
dt$id <- as.numeric(gsub("^.*fqs_header/.*_si(9095|9597|9798|9899|99100)([0-9]+).txt", "\\1", dt$file)) #nolint
dt$depth <- as.numeric(gsub("^.*fqs_header/.*_si(9095|9597|9798|9899|99100)([0-9]+).txt", "\\2", dt$file)) #nolint
dt <- dt[, !"file"]
# everthing before the first whitespace
dt$seqid <- sub(" .*$", "", dt$V1)
dt$seqid <- sub("^@", "", dt$seqid)
dt$len <- as.numeric(gsub("^.*\ length=([0-9]+)\ .*", "\\1", dt$V1))
dt$eflen <- as.numeric(gsub("^.*error-free_length=([0-9]+).*", "\\1", dt$V1))
dt$id2ref <- as.numeric(gsub("^.*read_identity=(.*)%", "\\1", dt$V1))

# separate V2 by ',' into multiple columns
dt$V2 <- gsub("[^ ]+ (.*) length=.*", "\\1", dt$V1)
# adaptor has no direction but simulated amplicon does
dt <- dt[, c("V3", "V4", "V5", "V6", "V7") := tstrsplit(V2, " ", fixed = TRUE)] #nolint
dt <- dt[, c("ref", "strand", "pos") := tstrsplit(V3, ",", fixed = TRUE)] #nolint
dt$type <- dt$ref
dt$type[dt$V4 == "chimera"] <- "chimera"
# ref_[0-9]+ -> ref_seq
dt$type <- gsub("^ref_[0-9]+", "ref_seq", dt$type)
dt <- dt[, !c("V1", "V3", "V4", "V5", "V6", "V7")]
df_ref <- as.data.frame(dt)
colnames(df_ref) <- c("benchmark", "profile", "id", "depth", "seqid", "len", "eflen", "id2ref", "info", "ref", "strand", "pos", "type") #nolint
df_ref$ref_strand <- paste0(df_ref$ref, df_ref$strand)

# cluster: P
fs <- list.files(
  path = "./data/simulate/clust",
  pattern = "*.tsv", full.names = TRUE, recursive = TRUE)
# grep meshclust_..
fs <- grep(pattern = '.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/.*meshclust_auto/meshclust/si.*', fs, value = TRUE) #nolint
dt <- rbindlist(sapply(fs, function(x) fread(x, header=FALSE, sep="\t", blank.lines.skip = TRUE), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt$benchmark <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\1", dt$file) #nolint
dt$profile <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\2", dt$file) #nolint
dt$clust <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/clust_compare/(.*)_auto/.*", "\\3", dt$file) #nolint
dt$meshclust <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/clust_compare/(.*)_auto/.*", "\\4", dt$file) #nolint
dt$id <- as.numeric(gsub("^.*/si(9095|9597|9798|9899|99100).*", "\\1", dt$file)) #nolint
dt$depth <- as.numeric(
  gsub("^.*/si(9095|9597|9798|9899|99100)([0-9]+).*", "\\2", dt$file)) #nolint
dt$base <- gsub("^.*/meshclust/(si.*).tsv", "\\1", dt$file) #nolint
cluster <- mapply(dt$base, dt$V1, FUN = function(x, y) paste0(x, "_", y)) #nolint
dt$V1 <- cluster
# rm ">" in V2
dt$V2 <- gsub("^>", "", dt$V2)
dt <- dt[, !c("file", "base")]
df_cls1 <- as.data.frame(dt)
colnames(df_cls1) <- c("cluster", "seqid", "est_id", "member", "benchmark", "profile", "clust", "meshclust", "id", "depth") #nolint

# pre-cluster: isONclust, umapcust, isONclust+umapclust
fs <- list.files(
  path = "./data/simulate/clust",
  pattern = "*.csv", full.names = TRUE, recursive = TRUE)
# grep read2cluster 
fs1 <- grep(pattern = '.*/isONclust_meshclust_auto/isONclust/read2cluster/si.*', fs, value = TRUE) #nolint
fs2 <- grep(pattern = '.*/umapclust_meshclust_auto/umapclust/read2cluster/si.*', fs, value = TRUE) #nolint
fs3 <- grep(pattern = '.*/isONclust_umapclust_meshclust_auto/umapclust/read2cluster/si.*', fs, value = TRUE) #nolint
fs <- c(fs1, fs2, fs3)
dt <- rbindlist(sapply(fs, function(x) fread(x, header=FALSE, sep="\t", blank.lines.skip = TRUE), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt$benchmark <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\1", dt$file) #nolint
dt$profile <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\2", dt$file) #nolint
dt$clust <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/clust_compare/(.*)_(auto|95)/(.*)/read2cluster/.*", "\\5", dt$file) #nolint
# isONclust+umapclust
index <- grep(".*/isONclust_umapclust_meshclust_auto/umapclust/read2cluster/", dt$file) #nolint
dt$clust[index] <- "isONclust+umapclust"

dt$meshclust <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/clust_compare/(.*)_(auto|95)/.*", "\\4", dt$file) #nolint
dt$id <- as.numeric(gsub("^.*/si(9095|9597|9798|9899|99100).*", "\\1", dt$file)) #nolint
dt$depth <- as.numeric(
  gsub("^.*/si(9095|9597|9798|9899|99100)([0-9]+).*", "\\2", dt$file)) #nolint
dt$base <- gsub("^.*/read2cluster/(si.*).csv", "\\1", dt$file) #nolint
dt <- dt[, !c("file")]
dt$est_id <- NA
dt$member <- NA
dt <- dt[ ,c("base", "V1", "est_id", "member", "benchmark", "profile", "clust", "meshclust", "id", "depth")] #nolint
df_cls2 <- as.data.frame(dt)
colnames(df_cls2) <- c("cluster", "seqid", "est_id", "member", "benchmark", "profile", "clust", "meshclust", "id", "depth") #nolint

# combine df_cls1, df_cls2
df_cls <- rbind.data.frame(df_cls1, df_cls2)
df_merge <- df_cls %>%
  full_join(df_ref, by = c("seqid", "benchmark", "profile", "id", "depth")) %>%
  group_by(benchmark, profile, meshclust, id, depth, cluster) %>%
  # number of reads supporting the cluster
  mutate(snreads = length(unique(seqid))) %>%
  ungroup() %>%
  # number of reads supporting the class
  group_by(benchmark, profile, meshclust, id, depth, cluster, ref_strand) %>%
  mutate(snreads2 = length(unique(seqid))) %>%
  # sort by nsreads, the most common class comes first
  arrange(desc(snreads2), ref_strand, .by_group = TRUE) %>%
  ungroup()

#########################
# snreads >= 10
df_merge <- df_merge %>%
  filter(snreads >= 10)

#https://stackoverflow.com/questions/9253843/r-clustering-purity-metric
purity <- function(clusters, classes) {
  sum(apply(table(classes, clusters), 2, max)) / length(clusters)
}
library(aricode)

df_comp0 <- df_merge %>%
  filter(! is.na(cluster)) %>%
  group_by(benchmark, profile, clust, meshclust, id, depth, cluster, snreads) %>% #nolint
  summarize(
    Purity = purity(as.integer(as.ordered(cluster)), as.integer(as.ordered(ref_strand))), #nolint
  ) %>%
  ungroup() %>%
  gather(key, value, -benchmark, -profile, -clust, -meshclust, -id, -depth, -cluster, -snreads) #nolint
library(ggprism)
library(RColorBrewer)
p0_cls <- df_comp0 %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS"))) %>%
  ggplot( 
    aes(
      x = snreads,
      y = value, 
      color = factor(
        clust,
        levels = c("isONclust", "isONclust_meshclust", "umapclust", "umapclust_meshclust", "isONclust+umapclust", "isONclust_umapclust_meshclust", "meshclust"), #nolint
        labels = c("isONclust","isONclust+", "umapclust", "umapclust+", "isONclust+umapclust", "isONclust+umapclust+", "meshclust"), #nolint
        ),
      shape = factor(
        profile,
        levels = c("even10", "skew10")
        ))) +  
  geom_point() +
  labs(
    x = "# of supporrting reads",
    y = "Purity",
    color = "Cluster mode",
    shape = "Profile"
   ) +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,2,3,4,6,7,8)]) +
  facet_grid(
    benchmark ~ factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ), scales = "free") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9) #nolint
    )
#p0_cls

# purity, NMI without considering strand 
# umapclust is based on canonical kmer, thus no sequnce orientation information
df_comp_nostrand <- df_merge %>%
  filter(! is.na(cluster)) %>%
  group_by(benchmark, profile, clust, meshclust, id, depth) %>%
  summarise(
    Size = length(unique(cluster)),
    Purity = purity(as.integer(as.ordered(cluster)), as.integer(as.ordered(ref))), #nolint
    NMI = NMI(as.integer(as.ordered(ref)), as.integer(as.ordered(cluster))), #nolint
  ) %>%
  gather(key, value, -benchmark, -profile, -clust, -meshclust, -id, -depth) #nolint

p1_cls_nostrand <- df_comp_nostrand %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS"))) %>%
  ggplot( 
    aes(
      x = factor(
        clust,
        levels = c("isONclust", "isONclust_meshclust", "umapclust", "umapclust_meshclust", "isONclust+umapclust", "isONclust_umapclust_meshclust", "meshclust"), #nolint
        labels = c("isONclust","isONclust+", "umapclust", "umapclust+", "isONclust+umapclust", "isONclust+umapclust+", "meshclust"), #nolint
        ),
      y = value, 
      color = factor(
        id, 
        levels = c(9095,9597,9798,9899,99100),
        labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%")), #nolint
      fill = factor(
        id, 
        levels = c(9095,9597,9798,9899,99100),
        labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
        ))) +
  geom_col(position = position_dodge2(width = 0.8, preserve = "single")) + #nolint
  geom_hline(data = data.frame(yint=c(10,1,1),key=c("Size", "Purity", "NMI")), aes(yintercept = yint), linetype = "dashed") + #nolint
  labs(
    x = "",
    y = "Quality of clustering",
    color = "% of Divergence",
    fill = "% of Divergence"
   ) +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  scale_fill_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  facet_grid(
    factor(
      key,
      levels = c("Size", "Purity", "NMI")) ~ benchmark + factor(
      profile,
      levels = c("even10", "skew10")), #nolint
    scales = "free", space = "free_x") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    strip.text = element_text(size = 8),
    )

# purity, NMI considering strand information
df_comp <- df_merge %>%
  filter(! is.na(cluster)) %>%
  group_by(benchmark, profile, clust, meshclust, id, depth) %>%
  summarise(
    Size = length(unique(cluster)),
    Purity = purity(as.integer(as.ordered(cluster)), as.integer(as.ordered(ref_strand))), #nolint
    NMI = NMI(as.integer(as.ordered(ref_strand)), as.integer(as.ordered(cluster))), #nolint
  ) %>%
  gather(key, value, -benchmark, -profile, -clust, -meshclust, -id, -depth) #nolint

p1_cls <- df_comp %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS"))) %>%
  ggplot( 
    aes(
      x = factor(
        clust,
        levels = c("isONclust", "isONclust_meshclust", "umapclust", "umapclust_meshclust", "isONclust+umapclust", "isONclust_umapclust_meshclust", "meshclust"), #nolint
        labels = c("isONclust","isONclust+", "umapclust", "umapclust+", "isONclust+umapclust", "isONclust+umapclust+", "meshclust"), #nolint
        ),
      y = value, 
      color = factor(
        id, 
        levels = c(9095,9597,9798,9899,99100),
        labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%")), #nolint
      fill = factor(
        id, 
        levels = c(9095,9597,9798,9899,99100),
        labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
        ))) +
  geom_col(position = position_dodge2(width = 0.8, preserve = "single")) + #nolint
  geom_hline(data = data.frame(yint=c(20,1,1),key=c("Size", "Purity", "NMI")), aes(yintercept = yint), linetype = "dashed") + #nolint
  labs(
    x = "",
    y = "Quality of clustering",
    color = "% of Divergence",
    fill = "% of Divergence"
   ) +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  scale_fill_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  facet_grid(
    factor(
      key,
      levels = c("Size", "Purity", "NMI")) ~ benchmark + factor(
      profile,
      levels = c("even10", "skew10")), #nolint
    scales = "free", space = "free_x") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    strip.text = element_text(size = 8),
    )
#p1_cls
# pct of clustered reads
df_clust <- df_merge %>%
  group_by(benchmark, profile, id, depth) %>%
  mutate(nreads = length(unique(seqid))) %>%
  filter(!is.na(cluster)) %>%
  group_by(benchmark, profile, clust, meshclust, id, depth, type) %>%
  reframe(pct = length(unique(seqid))/nreads * 100)  %>%
  # rm duplicate lines
  distinct()

p1_clust <- df_clust %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS"))) %>%
  ggplot( 
    aes(
      x = factor(
        clust,
        levels = c("isONclust", "isONclust_meshclust", "umapclust", "umapclust_meshclust", "isONclust+umapclust", "isONclust_umapclust_meshclust", "meshclust"), #nolint
        labels = c("isONclust","isONclust+", "umapclust", "umapclust+", "isONclust+umapclust", "isONclust+umapclust+", "meshclust"), #nolint
        ),
      y = pct, 
        )) +
  geom_bar(stat = "identity", aes(fill = type)) +
  labs(
    x = "",
    y = "% of clustered reads",
    fill = "Type"
   ) +
  scale_fill_manual(values = brewer.pal(4, "Spectral")[c(1,4)]) +
  facet_grid(
    factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ) ~ benchmark + profile,
      scales = "free_x", space = "free_x") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    # axis font angle
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    strip.text = element_text(size = 8),
    ) 
#p1_clust
p1_clustid <- df_merge %>%
  filter(! is.na(cluster)) %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS"))) %>%
  filter(!is.na(cluster)) %>%
  ggplot( 
    aes(
      x = factor(
        clust,
        levels = c("isONclust", "isONclust_meshclust", "umapclust", "umapclust_meshclust", "isONclust+umapclust", "isONclust_umapclust_meshclust", "meshclust"), #nolint
        labels = c("isONclust","isONclust+", "umapclust", "umapclust+", "isONclust+umapclust", "isONclust+umapclust+", "meshclust"), #nolint
        ),
      y = id2ref, 
        )) +
  geom_boxplot(
    aes(fill = type),
    position = position_dodge(preserve = "single"),
    outlier.size = 0.1,
    ) +
  labs(
    x = "",
    y = "% of identity relative to reference",
    fill = "Type"
   ) +
  scale_fill_manual(values = brewer.pal(4, "Spectral")[c(1,4)]) +
  facet_grid(
    factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ) ~ benchmark + profile,
      scales = "free_x", space = "free_x") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    # axis font angle
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    strip.text = element_text(size = 8),
    ) 
#p1_clustid
dir.create("./figure/simulate/clust_mode", recursive = TRUE, showWarnings = FALSE) #nolint
ggsave("./figure/simulate/clust_mode/p0_cls_point.png", p0_cls, width = 12, height = 10) #nolint
ggsave("./figure/simulate/clust_mode/p1_cls_bar_nostrand.png", p1_cls_nostrand, width = 12, height = 8) #nolint
ggsave("./figure/simulate/clust_mode/p1_cls_bar.png", p1_cls, width = 12, height = 8) #nolint
ggsave("./figure/simulate/clust_mode/p1_clust_pct.png", p1_clust, width = 12, height = 12) #nolint
ggsave("./figure/simulate/clust_mode/p1_clustid_identity.png", p1_clustid, width = 12, height = 12) #nolint
```

```{r, recall_loss}
# recall, percentage of reads correctly being grouped into one cluster
# loss, percentage of reads not being grouped into any cluster
library(data.table)
library(tidyverse)
# fqs_header: TP
fs <- list.files(
  path = "./data/simulate/consensus/fqs_header",
  pattern = "*[0-9].txt", full.names = TRUE, recursive = TRUE)
dt <- rbindlist(sapply(fs, function(x) fread(x, header=FALSE, sep=""), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt$benchmark <- gsub("^.*fqs_header/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)_(even10|skew10)_si.*.txt", "\\1", dt$file) #nolint
dt$profile <- gsub("^.*fqs_header/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)_(even10|skew10)_si.*.txt", "\\2", dt$file) #nolint
dt$id <- as.numeric(gsub("^.*fqs_header/.*_si(9095|9597|9798|9899|99100)([0-9]+).txt", "\\1", dt$file)) #nolint
dt$depth <- as.numeric(gsub("^.*fqs_header/.*_si(9095|9597|9798|9899|99100)([0-9]+).txt", "\\2", dt$file)) #nolint
dt <- dt[, !"file"]
# everthing before the first whitespace
dt$seqid <- sub(" .*$", "", dt$V1)
dt$seqid <- sub("^@", "", dt$seqid)
dt$len <- as.numeric(gsub("^.*\ length=([0-9]+)\ .*", "\\1", dt$V1))
dt$eflen <- as.numeric(gsub("^.*error-free_length=([0-9]+).*", "\\1", dt$V1))
dt$id2ref <- as.numeric(gsub("^.*read_identity=(.*)%", "\\1", dt$V1))

# separate V2 by ',' into multiple columns
dt$V2 <- gsub("[^ ]+ (.*) length=.*", "\\1", dt$V1)
# adaptor has no direction, but simulated amplicon does
dt <- dt[, c("V3", "V4", "V5", "V6", "V7") := tstrsplit(V2, " ", fixed = TRUE)] #nolint
dt <- dt[, c("ref", "strand", "pos") := tstrsplit(V3, ",", fixed = TRUE)] #nolint
dt$type <- dt$ref
dt$type[dt$V4 == "chimera"] <- "chimera"
# ref_[0-9]+ -> ref_seq
dt$type <- gsub("^ref_[0-9]+", "ref_seq", dt$type)
dt <- dt[, !c("V1", "V3", "V4", "V5", "V6", "V7")]
df_ref <- as.data.frame(dt)
colnames(df_ref) <- c("benchmark", "profile", "id", "depth", "seqid", "len", "eflen", "id2ref", "info", "ref", "strand", "pos", "type") #nolint
df_ref$ref_strand <- paste0(df_ref$ref, df_ref$strand)

# cluster: P
fs <- list.files(
  path = "./data/simulate/consensus",
  pattern = "*.csv", full.names = TRUE, recursive = TRUE)
dt <- rbindlist(sapply(fs, function(x) fread(x, header=FALSE, sep="\t", blank.lines.skip = TRUE), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt$benchmark <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\1", dt$file) #nolint
dt$profile <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\2", dt$file) #nolint
dt$consensus <- gsub("^.*(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/(kmerCon|miniCon|isoCon)/.*", "\\3", dt$file) #nolint
dt$id <- as.numeric(gsub("^.*/si(9095|9597|9798|9899|99100).*", "\\1", dt$file)) #nolint
dt$depth <- as.numeric(
  gsub("^.*/si(9095|9597|9798|9899|99100)([0-9]+).*", "\\2", dt$file)) #nolint
dt$cls <- gsub("^.*/clusters/si[0-9]+_(.*).csv", "\\1", dt$file)
dt <- dt[, !"file"]
df_cls <- as.data.frame(dt)
colnames(df_cls) <- c("seqid", "benchmark", "profile", "consensus", "id", "depth", "cls") #nolint

df_merge <- df_cls %>%
  full_join(df_ref, by = c("seqid", "benchmark", "profile", "id", "depth")) %>%
  group_by(benchmark, profile, id, depth, consensus, cls) %>%
  # number of reads supporting the cluster
  mutate(snreads = length(unique(seqid))) %>%
  ungroup() %>%
  # number of reads supporting the class
  group_by(benchmark, profile, id, depth, consensus, cls, ref_strand) %>%
  mutate(snreads2 = length(unique(seqid))) %>%
  # sort by nsreads, the most common class comes first
  arrange(desc(snreads2), ref_strand, .by_group = TRUE) %>%
  ungroup()

#https://stackoverflow.com/questions/9253843/r-clustering-purity-metric
purity <- function(clusters, classes) {
  sum(apply(table(classes, clusters), 2, max)) / length(clusters)
}
library(aricode)

df_comp0 <- df_merge %>%
  filter(! is.na(cls)) %>%
  group_by(benchmark, profile, id, depth, consensus, cls, snreads) %>%
  summarize(
    Purity = purity(as.integer(as.ordered(cls)), as.integer(as.ordered(ref_strand))), #nolint
  ) %>%
  ungroup() %>%
  gather(key, value, -benchmark, -profile, -id, -depth, -consensus, -cls, -snreads) #nolint
library(ggprism)
library(RColorBrewer)
p0_cls <- df_comp0 %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"), #nolint
      )) %>%
  ggplot(aes(x = snreads, y = value, color = consensus, shape = profile)) +
  geom_point() +
  labs(
    x = "# of supporrting reads",
    y = "Purity",
    color = "% of Divergence",
    shape = "Simulation profile"
   ) +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,8)]) +
  facet_grid( benchmark ~ id, scales = "free") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    )
#p0_cls
df_comp <- df_merge %>%
  filter(! is.na(cls)) %>%
  group_by(benchmark, profile, id, depth, consensus) %>%
  summarise(
    Size = length(unique(cls)),
    Purity = purity(as.integer(as.ordered(cls)), as.integer(as.ordered(ref_strand))), #nolint
    NMI = NMI(as.integer(as.ordered(ref_strand)), as.integer(as.ordered(cls))),
  ) %>%
  gather(key, value, -benchmark, -profile, -id, -depth, -consensus)

p1_cls <- df_comp %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  filter(key == "Purity") %>%
  ggplot(aes(x = depth, y = value, color = id, shape = profile)) +
  geom_point() +
  geom_line() +
  labs(
    x = "Sequencing depth",
    y = "Quality of clustering",
    color = "% of Divergence",
    shape = "Simulation profile"
   ) +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  scale_x_continuous(
    trans = scales::log2_trans(),
    breaks = c(10,20,40,60,80,100,200,400,800),
    labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7") #nolint
    ) +
  facet_grid(key + consensus ~ benchmark, scales = "free_y") +
  geom_hline(data = data.frame(yint=c(1),key=c("Purity")), aes(yintercept = yint), linetype = "dashed") + #nolint
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    )
#p1_cls
p1_cls_size_nmi <- df_comp %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  filter(key %in% c("Size", "NMI")) %>%
  ggplot(aes(x = depth, y = value, color = id, shape = profile)) +
  geom_point() +
  geom_line() +
  labs(
    x = "Sequencing depth",
    y = "Quality of clustering",
    color = "% of Divergence",
    shape = "Simulation profile"
   ) +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  scale_x_continuous(
    trans = scales::log2_trans(),
    breaks = c(10,20,40,60,80,100,200,400,800),
    labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7") #nolint
    ) +
  facet_grid(factor(key, levels = c("Size", "NMI")) + consensus ~ benchmark, scales = "free_y") + #nolint
  geom_hline(data = data.frame(yint=c(20,1),key=c("Size", "NMI")), aes(yintercept = yint), linetype = "dashed") + #nolint
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9) #nolint
    )
#p1_cls_size_nmi

# pct: cluster redas
df_pct_clust <- df_merge %>%
  group_by(benchmark, profile, id, depth) %>%
  mutate(nreads = length(unique(seqid))) %>%
  filter(! is.na(consensus)) %>%
  group_by(benchmark, profile, id, depth, consensus, type) %>%
  reframe(pct = length(unique(seqid)) / nreads * 100)  %>%
  # rm duplicate lines
  distinct()

p1_pct_clust <- df_pct_clust %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  ggplot( 
    aes(
      x = factor(
        depth,
        levels = c(10,20,40,60,80,100,200,400,800),
        labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7") #nolint
        ), 
      y = pct, 
        )) +
  geom_bar(stat = "identity", aes(fill = type)) +
  labs(
    x = "Sequencing depth",
    y = "% of reads for consensus calling",
    fill = "Type"
   ) +
  scale_fill_manual(values = brewer.pal(4, "Spectral")[c(1, 4)]) +
  facet_grid(id + profile ~ benchmark + consensus) +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    strip.text = element_text(size = 8),
    )  
#p1_pct_clust
#ggsave("./figure/simulate/consensus/p1_clust.png", p1_pct_clust, width = 18, height = 12) #nolint
p1_clustid <- df_merge %>%
  filter(! is.na(cls)) %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  ggplot( 
    aes(
      x = factor(
        depth,
        levels = c(10,20,40,60,80,100,200,400,800),
        labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7") #nolint
        ), 
      y = id2ref, 
        )) +
  geom_boxplot(
    aes(fill = type),
    position = position_dodge(preserve = "single"),
    outlier.size = 0.1,
    ) +
  labs(
    x = "Sequencing depth",
    y = "% of identity relative to reference",
    fill = "Type"
   ) +
  scale_fill_manual(values = brewer.pal(4, "Spectral")[c(1,4)]) +
  facet_grid(id + profile ~ benchmark + consensus) +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9), #nolint
    strip.text = element_text(size = 8),
    )
#p1_clustid 
```

```{r, clean}
library(data.table)
library(tidyverse)
fs_all <- list.files(
  path = "./data/simulate/consensus",
  pattern = "*.paf", full.names = TRUE, recursive = TRUE)
fs <- grep(pattern = '.*/cut/.*.paf', fs_all, value = TRUE)

dt <- rbindlist(sapply(fs, function(x) fread(x, header=FALSE), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt$benchmark <- gsub("^.*simulate/consensus/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\1", dt$file) #nolint
dt$profile <- gsub("^.*simulate/consensus/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\2", dt$file) #nolint
dt$id <- as.numeric(
  gsub("^.*/si(9095|9597|9798|9899|99100).*", "\\1", dt$file)) #nolint
dt$depth <- as.numeric(
  gsub("^.*/si(9095|9597|9798|9899|99100)([0-9]+).*", #nolint
  "\\2", dt$file))
dt$alignment <- gsub("^.*/si[0-9]+_(asm[0-9]+).*", "\\1", dt$file)
dt$consensus <- gsub("^.*(even10|skew10)/(.*)/repseqs_refalign/cut/.*", "\\2", dt$file) #nolint
dt <- dt[, !"file"]
df <- as.data.frame(dt)
# https://github.com/lh3/miniasm/blob/master/PAF.md
colnames(df) <- c("qseq", "qlen", "qstart", "qend", "strand",
 "tseq", "tlen", "tstart", "tend", "matches", "alnlen", "mapq",
  "benchmark", "profile", "id", "depth", "asm", "consensus")

# df2: number of representative sequences
fs2 <-
 list.files(
  path = "./data/simulate/consensus",
  pattern = "*repfasta_lines.txt", full.names = TRUE, recursive = TRUE)
dt2 <- rbindlist(sapply(fs2, fread, simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt2$benchmark <- gsub("^.*simulate/consensus/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\1", dt2$file) #nolint
dt2$profile <- gsub("^.*simulate/consensus/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\2", dt2$file) #nolint
dt2$id <- as.numeric(gsub("^.*/si(9095|9597|9798|9899|99100).*", "\\1", dt2$V2)) #nolint
dt2$depth <- as.numeric(gsub("^.*/si(9095|9597|9798|9899|99100)([0-9]+).*", "\\2", dt2$V2)) #nolint
dt2$consensus <- gsub("^.*10/(.*Con.*)/repseqs_split.*", "\\1", dt2$V2)
dt2 <- dt2[, !c("file", "V2")]
df2 <- as.data.frame(dt2)
# fasta: two lines per read
df2$V1 <- df2$V1 / 2
colnames(df2) <- c("nseqs", "benchmark", "profile", "id", "depth", "consensus")

# df3: number of simulated reads
fs3 <- list.files(
  path = "./data/simulate/consensus",
  pattern = "*sifastq_lines.txt", full.names = TRUE, recursive = TRUE)
dt3 <- rbindlist(sapply(fs3, function(x) fread(x, header=FALSE), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt3$benchmark <- gsub("^.*simulate/consensus/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\1", dt3$file) #nolint
dt3$profile <- gsub("^.*simulate/consensus/(nanopore2020|nanopore2023|nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\2", dt3$file) #nolint
dt3$id <- as.numeric(gsub("^.*/si(9095|9597|9798|9899|99100).*", "\\1", dt3$V2)) #nolint
dt3$depth <- as.numeric(gsub("^.*/si(9095|9597|9798|9899|99100)([0-9]+).*", "\\2", dt3$V2)) #nolint
dt3$consensus <- gsub("^.*10/(.*Con.*)/sifastq_lines.txt", "\\1", dt3$file)
dt3 <- dt3[, !c("file","V2")]
df3 <- as.data.frame(dt3)
# fastq: four lines per read
df3$V1 <- df3$V1 / 4
colnames(df3) <- c("reads", "benchmark", "profile", "id", "depth", "consensus")
#left_join
df23 <- df2 %>%
  left_join(df3, by = c("benchmark", "profile", "id", "depth", "consensus")) 
```

```{r, visual-dt2, eval = FALSE}
library(ggplot2)
library(RColorBrewer)
library(ggprism)
library(forcats)
p1 <- df23 %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  ggplot( 
    aes(x = depth, y = nseqs, color = id, shape = profile)) +
  geom_point() +
  geom_line() +
  labs(
    x = "Sequencing depth",
    y = "# of polished sequences",
    color = "% of Divergence",
    shape = "Simulation profile"
   ) +
  # add line y = 20, 10 refs in two strands
  geom_hline(yintercept = 20, linetype = "dashed") +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  scale_y_continuous(breaks = c(0, 10, 20, 50, 100, 150)) +
  scale_x_continuous(
    trans = scales::log2_trans(),
    breaks = c(10,20,40,60,80,100,200,400,800),
    labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7"), #nolint
    ) +
  facet_grid(consensus ~ benchmark, scales = "free_y") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1))
#p1
```

```{r, visual-dt1}
df$asm <- factor(df$asm, levels = c("asm5", "asm10", "asm20"))
# grouped stats per benchmark
# !!!strand drep rev complement: representative sequences
df123 <- df %>%
  group_by(benchmark, profile, id, depth, asm, consensus) %>%
  # mapq = 60: unique mapping
  # https://github.com/lh3/minimap2/issues/447
  # http://www.acgt.me/blog/2014/12/16/understanding-mapq-scores-in-sam-files-does-37-42#:~:text=MAPQ%3A%20MAPping%20Quality.,mapping%20quality%20is%20not%20available.
  #filter(mapq >= 60) %>%
  summarise(
    tn = n_distinct(tseq),
    qn = n_distinct(qseq),
  ) %>%
  # join df2
  left_join(df23, by = c("benchmark", "profile","id", "depth", "consensus")) %>% #nolint
  mutate(
    qper = qn / nseqs * 100,
    tper = tn / 10 * 100,
  )

p2 <- df123 %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  ggplot(
  aes(x = depth, y = tn, color = id, shape = profile)) + 
  geom_point() +
  geom_line() +
  labs(
    x = "Sequencing depth",
    y = "# of recovered target sequences",
    color = "% of Divergence",
    shape = "Simulation profile"
   ) +
  # add line y = 10, 10 refs in two strands, only report number of refs in alignment
  geom_hline(yintercept = 10, linetype = "dashed") +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  scale_x_continuous(
    trans = scales::log2_trans(),
    breaks = c(10,20,40,60,80,100,200,400,800),
    labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7"), #nolint
    ) +
  facet_grid(asm + consensus ~ benchmark, scales = "free_y") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9) #nolint
    )
#p2

p2_asm5 <- df123 %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  filter(asm == "asm5") %>%
  ggplot(
  aes(x = depth, y = tn, color = id, shape = profile)) + 
  geom_point() +
  geom_line() +
  labs(
    x = "Sequencing depth",
    y = "# of recovered target sequences",
    color = "% of Divergence",
    shape = "Simulation profile"
   ) +
  # add line y = 10, 10 refs in two strands, only report number of refs in alignment
  geom_hline(yintercept = 10, linetype = "dashed") +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  scale_x_continuous(
    trans = scales::log2_trans(),
    breaks = c(10,20,40,60,80,100,200,400,800),
    labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7"), #nolint
    ) +
  facet_grid(asm + consensus ~ benchmark, scales = "free_y") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9) #nolint
    )
#p2_asm5

p3 <- df123 %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  ggplot(
  aes(x = depth, y = qper, color = id, shape = profile)) + 
  geom_point() +
  geom_line() +
  labs(
    x = "Sequencing depth",
    y = "% of aligned sequences",
    color = "% of Divergence",
    shape = "Simulation profile"
   ) +
  # add line y = 20, 10 refs in two strands
  geom_hline(yintercept = 20, linetype = "dashed") +
  scale_color_manual(values = brewer.pal(8, "Spectral")[c(1,3,4,7,8)]) +
  scale_x_continuous(
    trans = scales::log2_trans(),
    breaks = c(10,20,40,60,80,100,200,400,800),
    labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7"), #nolint
    ) +
  facet_grid(asm + consensus ~ benchmark, scales = "free_y") +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 9) #nolint
    )
#p3

# alignment accuracy
# % of matches
df$qcov <- df$matches / df$qlen * 100
df$tcov <- df$matches / df$tlen * 100
# % of matches, blast-like identity
df$alnid <- df$matches / (df$alnlen) * 100

# blast-like identity
library(ggbeeswarm)
p4_alnid <- df %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  filter(asm == "asm10") %>%
  ggplot(aes(
    x = factor(
      depth,
      levels = c(10,20,40,60,80,100,200,400,800),
      labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7") #nolint
      ), 
    y = 100 - alnid, color = mapq)) +
  geom_quasirandom(aes(shape = profile)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = 0.1, linetype = "dashed") +
  geom_hline(yintercept = 1, linetype = "dashed") +
  labs(
    x = "",
    y = "% of errors in alignment",
    color = "MAPQ",
    shape = "Simulation profile"
   ) +
  scale_color_distiller(palette = "Spectral") +
  scale_y_continuous(
    trans = scales::sqrt_trans(),
    breaks = c(0, 0.1, 1, 2, 3, 5),
    ) +
  facet_grid(id ~ benchmark + consensus) +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 8), #nolint
    strip.text = element_text(size = 8),
    )
#p4_alnid 

# qcov
p4_qcov <- df %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  filter(asm == "asm10") %>%
  ggplot(aes(
    x = factor(
      depth,
      levels = c(10,20,40,60,80,100,200,400,800),
      labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7") #nolint
      ), 
    y = 100 - qcov, color = mapq)) +
  geom_quasirandom(aes(shape = profile)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = 0.1, linetype = "dashed") +
  geom_hline(yintercept = 1, linetype = "dashed") +
  labs(
    x = "",
    y = "% of unaligned regions in the corrected sequences",
    color = "MAPQ",
    shape = "Simulation profile"
   ) +
  scale_color_distiller(palette = "Spectral") +
  scale_y_continuous(
    trans = scales::sqrt_trans(),
    breaks = c(0, 0.1, 1, 2, 3, 5),
    ) +
  facet_grid(id ~ benchmark + consensus) +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 8), #nolint
    strip.text = element_text(size = 8),
    )
#p4_qcov 

# tcov
p4_tcov <- df %>%
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023", "nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT 2020", "ONT 2023", "ONT Duplex", "PacBio CCS")
    ),
    id = factor(
      id, 
      levels = c(9095,9597,9798,9899,99100),
      labels = c("5~10%", "3~5%", "2~3%", "1~2%", "0~1%") #nolint
      ),
    consensus = factor(
      consensus,
      levels = c("kmerCon", "miniCon", "isoCon"))) %>%
  filter(depth !=10 & depth != 800) %>%
  filter(asm == "asm10") %>%
  ggplot(aes(
    x = factor(
      depth,
      levels = c(10,20,40,60,80,100,200,400,800),
      labels = c("10\u00D7", "20\u00D7", "40\u00D7", "60\u00D7", "80\u00D7", "100\u00D7", "200\u00D7", "400\u00D7", "800\u00D7") #nolint
      ), 
    y = 100 - tcov, color = mapq)) +
  geom_quasirandom(aes(shape = profile)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_hline(yintercept = 0.1, linetype = "dashed") +
  geom_hline(yintercept = 1, linetype = "dashed") +
  labs(
    x = "",
    y = "% of unaligned regions in the reference sequences",
    color = "MAPQ",
    shape = "Simulation profile"
   ) +
  scale_color_distiller(palette = "Spectral") +
  scale_y_continuous(
    trans = scales::sqrt_trans(),
    breaks = c(0, 0.1, 1, 2, 3, 5),
    ) +
  facet_grid(id ~ benchmark + consensus) +
  theme_prism() +
  theme(
    legend.title = element_text(),
    axis.text.x.bottom = element_text(angle = 90, vjust = 0.5, hjust = 1, size = 8), #nolint
    strip.text = element_text(size = 8),
  )
#p4_tcov 
```

```{r, summary_table}
# umapclust_mauto 
# sd for qcov, tcov, alnid, and pct of alnid > 99%
sum_table1 <- df %>%
  filter(depth !=10 & depth != 800) %>%
  filter(asm == "asm10" & benchmark %in% c("nanopore2023duplex", "pacbio2016ccs")) %>% #nolint
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT Duplex", "PacBio CCS")
    )) %>%  
  group_by(benchmark, profile, id, depth, asm, consensus) %>%
  summarise(
    `qcov:mean` = mean(qcov),
    `qcov:sd` = sd(qcov),
    `tcov:mean` = mean(tcov),
    `tcov:sd` = sd(tcov),
    `alnid:mean` = mean(alnid),
    `alnid:sd` = sd(alnid),
    n = n(),
    `pct:alnid >=99` = sum(alnid >= 99) / n() * 100,
  ) %>%
  arrange(`pct:alnid >=99`) %>%
  # two digits after decimal point
  mutate(
    `qcov:mean` = round(`qcov:mean`, 2),
    `qcov:sd` = round(`qcov:sd`, 2),
    `tcov:mean` = round(`tcov:mean`, 2),
    `tcov:sd` = round(`tcov:sd`, 2),
    `alnid:mean` = round(`alnid:mean`, 2),
    `alnid:sd` = round(`alnid:sd`, 2),
    `pct:alnid >=99` = round(`pct:alnid >=99`, 2),
  )

sum_table11 <- df %>%
  filter(depth !=10 & depth != 800) %>%
  filter(asm == "asm10" & benchmark %in% c("nanopore2020", "nanopore2023")) %>% #nolint
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2020", "nanopore2023"), #nolint
      labels = c("ONT 2020", "ONT 2023")
    )) %>%  
  group_by(benchmark, profile, id, depth, asm, consensus) %>%
  summarise(
    `qcov:mean` = mean(qcov),
    `qcov:sd` = sd(qcov),
    `tcov:mean` = mean(tcov),
    `tcov:sd` = sd(tcov),
    `alnid:mean` = mean(alnid),
    `alnid:sd` = sd(alnid),
    n = n(),
    `pct:alnid >=99` = sum(alnid >= 99) / n() * 100,
  ) %>%
  arrange(`pct:alnid >=99`) %>%
  # two digits after decimal point
  mutate(
    `qcov:mean` = round(`qcov:mean`, 2),
    `qcov:sd` = round(`qcov:sd`, 2),
    `tcov:mean` = round(`tcov:mean`, 2),
    `tcov:sd` = round(`tcov:sd`, 2),
    `alnid:mean` = round(`alnid:mean`, 2),
    `alnid:sd` = round(`alnid:sd`, 2),
    `pct:alnid >=99` = round(`pct:alnid >=99`, 2),
  )
sum_table12 <- rbind.data.frame(sum_table1, sum_table11)

# isONclust_umapclust_m5
# nanopore2023duplex pacbio2016ccs
fs_all <- list.files(
  path = "./data/simulate/isONclust_umapclust_m5",
  pattern = "*.paf", full.names = TRUE, recursive = TRUE)
fs <- grep(pattern = '.*/cut/.*.paf', fs_all, value = TRUE)

dt <- rbindlist(sapply(fs, function(x) fread(x, header=FALSE), simplify = FALSE), idcol = "file", use.names = FALSE) #nolint
dt$benchmark <- gsub("^.*simulate/isONclust_umapclust_m5/(nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\1", dt$file) #nolint
dt$profile <- gsub("^.*simulate/isONclust_umapclust_m5/(nanopore2023duplex|pacbio2016ccs)/(even10|skew10)/.*", "\\2", dt$file) #nolint
dt$id <- as.numeric(
  gsub("^.*/si(9095|9597|9798|9899|99100).*", "\\1", dt$file)) #nolint
dt$depth <- as.numeric(
  gsub("^.*/si(9095|9597|9798|9899|99100)([0-9]+).*", #nolint
  "\\2", dt$file))
dt$alignment <- gsub("^.*/si[0-9]+_(asm[0-9]+).*", "\\1", dt$file)
dt$consensus <- gsub("^.*(even10|skew10)/(.*)/repseqs_refalign/cut/.*", "\\2", dt$file) #nolint
dt <- dt[, !"file"]
df2 <- as.data.frame(dt)
# https://github.com/lh3/miniasm/blob/master/PAF.md
colnames(df2) <- c("qseq", "qlen", "qstart", "qend", "strand",
 "tseq", "tlen", "tstart", "tend", "matches", "alnlen", "mapq",
  "benchmark", "profile", "id", "depth", "asm", "consensus")
# alignment accuracy
# % of matches
df2$qcov <- df2$matches / df2$qlen * 100
df2$tcov <- df2$matches / df2$tlen * 100
# % of matches, blast-like identity
df2$alnid <- df2$matches / (df2$alnlen) * 100

sum_table2 <- df2 %>%
  filter(depth !=10 & depth != 800) %>%
  filter(asm == "asm10" & benchmark %in% c("nanopore2023duplex", "pacbio2016ccs")) %>% #nolint
  mutate(
    benchmark = factor(
      benchmark, 
      levels = c("nanopore2023duplex", "pacbio2016ccs"), #nolint
      labels = c("ONT Duplex", "PacBio CCS")
    )) %>%  
  group_by(benchmark, profile, id, depth, asm, consensus) %>%
  summarise(
    `qcov:mean` = mean(qcov),
    `qcov:sd` = sd(qcov),
    `tcov:mean` = mean(tcov),
    `tcov:sd` = sd(tcov),
    `alnid:mean` = mean(alnid),
    `alnid:sd` = sd(alnid),
    n = n(),
    `pct:alnid >=99` = sum(alnid >= 99) / n() * 100, #nolint
  ) %>%
  arrange(`pct:alnid >=99`) %>%
  # two digits after decimal point
  mutate(
    `qcov:mean` = round(`qcov:mean`, 2),
    `qcov:sd` = round(`qcov:sd`, 2),
    `tcov:mean` = round(`tcov:mean`, 2),
    `tcov:sd` = round(`tcov:sd`, 2),
    `alnid:mean` = round(`alnid:mean`, 2),
    `alnid:sd` = round(`alnid:sd`, 2),
    `pct:alnid >=99` = round(`pct:alnid >=99`, 2),
  )

dir.create("table/simulate", showWarnings = FALSE, recursive = TRUE)
write.csv(sum_table12, "table/simulate/sum_table1.csv", row.names = FALSE, quote = FALSE) #nolint
write.csv(sum_table2, "table/simulate/sum_table2.csv", row.names = FALSE, quote = FALSE) #nolint
```

```{r, pngs}
dir.create("figure/simulate/consensus", showWarnings = FALSE, recursive = TRUE)
library(patchwork)
p1 <- p1_cls / p2_asm5 +
 plot_annotation(tag_levels = "a") +
 plot_layout(guides = "collect")

ps <- list(p1_cls_size_nmi, p2, p3)
pnames <- c("p1_cls_size_nmi","p2_refcov", "p3_tseqsnum")
ps1 <- list(p4_alnid, p4_qcov, p4_tcov)
pnames1 <- c("p4_alnid", "p4_qcov", "p4_tcov")
ggsave2 <- function(x, y, width = 14, height = 12) {
  ggsave(
    paste0("figure/simulate/consensus/", x, ".png"),
    y,
    width = width,
    height = height,
    )
}
ggsave2("p1_cls", p1_cls, width = 12, height = 4)
ggsave2("p1", p1, width = 12, height = 10)
mapply(ggsave2, pnames, ps, 12, 10)
mapply(ggsave2, pnames1, ps1, 14, 12)
ps2 <- list(p1_pct_clust, p1_clustid, p0_cls)
pnames2 <- c("p1_pct_clust", "p1_clustid", "p0_cls")
mapply(ggsave2, pnames2, ps2, 14, 10)
```
